### PR TITLE
fix timings

### PR DESCRIPTION
**Pull request by bloodearnest**
- fix: ensure int timestamp in negative duration handling
- refactor: subsume set_state into set_code
- feat: make run loop handle synchronous operations

### CODE REVIEW

File: tests/test_run.py [hunk 1]
```diff
@@ -90,6 +82,80 @@ def test_handle_job_full_execution(synchronous_transitions, db, freezer):
...
 
 
+def test_handle_job_full_execution_synchronous(db, freezer):
+    # move to a whole second boundary for easier timestamp maths
+    freezer.move_to("2022-01-01T12:34:56")
+
+    api = StubExecutorAPI()
+    api.synchronous_transitions = [ExecutorState.PREPARING, ExecutorState.FINALIZING]
+
+    start = int(time.time() * 1e9)
```
Comment (madwort) [inline tests/test_run.py:92]: any reason not to use `time_ns`?
```suggestion
    start = int(time.time_ns())
```

File: jobrunner/run.py [hunk 2]
```diff
@@ -550,20 +533,42 @@ def set_code(job, code, message, error=None, results=None, timestamp_ns=None, **
...
-
-    if job.status_code_updated_at > timestamp_ns:
-        # we somehow have a negative duration, which honeycomb does funny things with.
-        # This can happen in tests, where things are fast, but we've seen it in production too.
-        log.warning(
-            f"negative state duration, clamping to 1ms ({job.status_code_updated_at} > {timestamp_ns})"
-        )
-        timestamp_ns = job.status_code_updated_at + 1e6  # set duration to 1ms
-        timestamp_s = int(timestamp_ns // 1e9)
+        timestamp_s = int(timestamp_ns / 1e9)
```
Comment (madwort) [inline jobrunner/run.py:536]: double-checking this should be `/` and not `//` 
```suggestion
        timestamp_s = int(timestamp_ns // 1e9)
```

**Review by madwort — APPROVED** [review]

File: tests/test_run.py (hunk 1 above)
Comment (bloodearnest) [reply tests/test_run.py:92]: nope, I'm slowly moveing to time_ns(), I didn't know about it when I originally wrote this.

File: jobrunner/run.py (hunk 2 above)
Comment (bloodearnest) [reply jobrunner/run.py:536]: yep, agreed

**Pull request merged**